crates/kornia-core-ops/src/ops.rs

```diff
@@ -1,78 +1,203 @@
+    /// assert_eq!(norm.shape, [1]);
+    /// assert_eq!(norm.as_slice(), [4.]);
+    /// ```
+    fn euclidean_norm(&self, dim: usize) -> Result<Tensor<T, N, A>, TensorOpsError>
+    where
+        T: std::ops::Add<Output = T> + Float;
+}
 
-    let mut out_shape = tensor.shape;
-    out_shape[dim] = 1;
+impl<T, const N: usize, A> TensorOps<T, N, A> for Tensor<T, N, A>
+where
+    T: SafeTensorType,
+    A: TensorAllocator + 'static,
+{
+    fn powf(&self, n: T) -> Tensor<T, N, A>
+    where
+        T: Float,
+    {
+        self.map(|x| x.powf(n))
+    }
 
-    let mut out_strides = tensor.strides;
-    if dim > 0 {
-        out_strides
-            .iter_mut()
-            .take(dim)
-            .for_each(|s| *s /= tensor.shape[dim]);
+    fn sqrt(&self) -> Tensor<T, N, A>
+    where
+        T: Float,
+    {
+        self.map(|x| x.sqrt())
     }
 
-    let numel: usize = out_shape.iter().product();
-    let mut data = vec![T::default(); numel];
-
-    for (i, v) in tensor.as_slice().iter().enumerate() {
-        let mut out_index = tensor.get_index_unchecked(i);
-        out_index[dim] = 0;
-        let out_offset = out_index
-            .iter()
-            .zip(out_strides.iter())
-            .fold(0, |acc, (&idx, &stride)| acc + idx * stride);
-        let agg = unsafe { data.get_unchecked_mut(out_offset) };
-        *agg = *agg + *v;
+    fn sum_elements(&self, dim: usize) -> Result<Tensor<T, N, A>, TensorOpsError>
+    where
+        T: std::ops::Add<Output = T>,
+    {
+        if dim >= N {
+            return Err(TensorOpsError::DimOutOfBounds(dim, N - 1));
+        }
+
+        let mut out_shape = self.shape;
+        out_shape[dim] = 1;
+
+        let mut out_strides = self.strides;
+        if dim > 0 {
+            out_strides
+                .iter_mut()
+                .take(dim)
+                .for_each(|s| *s /= self.shape[dim]);
+        }
+
+        let numel: usize = out_shape.iter().product();
+        let mut data = vec![T::default(); numel];
+
+        for (i, v) in self.as_slice().iter().enumerate() {
+            let mut out_index = self.get_index_unchecked(i);
+            out_index[dim] = 0;
+            let out_offset = out_index
+                .iter()
+                .zip(out_strides.iter())
+                .fold(0, |acc, (&idx, &stride)| acc + idx * stride);
+            let agg = unsafe { data.get_unchecked_mut(out_offset) };
+            *agg = *agg + *v;
+        }
+
+        let storage = TensorStorage::from_vec(data, self.storage.alloc().clone());
+
+        Ok(Tensor {
+            storage,
+            shape: out_shape,
+            strides: out_strides,
+        })
     }
 
-    let storage = TensorStorage::from_vec(data, tensor.storage.alloc().clone());
+    fn pnorm(&self, p: T, dim: usize) -> Result<Tensor<T, N, A>, TensorOpsError>
+    where
+        T: std::ops::Add<Output = T> + Float,
+    {
+        let p_inv = T::one() / p;
+        Ok(self.powf(p).sum_elements(dim)?.powf(p_inv))
```

Comment:
Should we provide a reduction argument?

Reply:
Do you mean to drop the dimension that the operation was run over?